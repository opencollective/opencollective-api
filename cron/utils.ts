import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
