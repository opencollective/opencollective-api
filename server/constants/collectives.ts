import config from 'config';

export enum CollectiveType {
  COLLECTIVE = 'COLLECTIVE',
  EVENT = 'EVENT',
  USER = 'USER',
  ORGANIZATION = 'ORGANIZATION',
  BOT = 'BOT',
  PROJECT = 'PROJECT',
  FUND = 'FUND',
  VENDOR = 'VENDOR',
}

export const CollectiveTypesList = Object.values(CollectiveType);

export const DEFAULT_BACKGROUND_IMG = `${config.host.website}/public/images/collectives/default-header-bg.jpg`;

export const defaultBackgroundImage = {
  COLLECTIVE: `${config.host.website}/static/images/defaultBackgroundImage.png`,
  USER: `${config.host.website}/static/images/defaultBackgroundImage-profile.svg`,
};
