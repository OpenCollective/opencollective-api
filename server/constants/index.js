import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, roles, math, channels, transactions };
