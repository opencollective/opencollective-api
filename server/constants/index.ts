import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import expenseTypes from './expense_type';
import frequencies from './frequencies';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, expenseTypes, roles, math, channels, transactions, frequencies };
