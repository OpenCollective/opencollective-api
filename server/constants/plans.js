const plans = {
  // Legacy Plans (automatically set for accounts created before 2020)
  'legacy-custom': {
    hostedCollectivesLimit: 99,
    addedFundsLimit: null,
    hostDashboard: true,
  },
  'legacy-large': {
    hostedCollectivesLimit: 24,
    addedFundsLimit: null,
    hostDashboard: true,
  },
  'legacy-medium': {
    hostedCollectivesLimit: 9,
    addedFundsLimit: null,
    hostDashboard: true,
  },
  'legacy-small': {
    hostedCollectivesLimit: 1,
    addedFundsLimit: null,
    hostDashboard: true,
  },
  // Plans (for customers from 2020)
  owned: {
    hostedCollectivesLimit: null,
    addedFundsLimit: null,
    hostDashboard: true,
  },
  custom: {
    hostedCollectivesLimit: 99,
    addedFundsLimit: null,
    hostDashboard: true,
    slug: 'network-host-plan',
    level: 4,
  },
  large: {
    hostedCollectivesLimit: 24,
    addedFundsLimit: null,
    hostDashboard: true,
    slug: 'large-host-plan',
    level: 3,
  },
  medium: {
    hostedCollectivesLimit: 9,
    addedFundsLimit: null,
    hostDashboard: true,
    slug: 'medium-host-plan',
    level: 2,
  },
  small: {
    hostedCollectivesLimit: 1,
    addedFundsLimit: null,
    hostDashboard: false,
    slug: 'small-host-plan',
    level: 1,
  },
  // Special plan for everyone without a plan
  default: {
    hostedCollectivesLimit: 1,
    addedFundsLimit: 100000, // in dollar cents
    hostDashboard: false,
  },
};

export const PLANS_COLLECTIVE_SLUG = 'opencollective';

export default plans;
