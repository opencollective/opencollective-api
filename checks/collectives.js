import '../server/env';

import { sequelize } from '../server/models';
// import models, { Op } from '../server/models';

const check = true;
// const fix = false;

async function checkDeletedUsers() {
  const results = await sequelize.query(
    `SELECT COUNT(*) as count
     FROM "Collectives" c
     LEFT JOIN "Users" u
     ON c."id" = u."CollectiveId"
     WHERE c."type" = 'USER'
     AND c."isIncognito" IS FALSE
     AND c."deletedAt" IS NULL
     AND (u."deletedAt" IS NOT NULL or u."id" IS NULL)`,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results[0].count > 0) {
    if (check) {
      throw new Error('No USER Collective without a matching User');
    }
  }
}

export async function checkCollectives() {
  await checkDeletedUsers();
}

if (!module.parent) {
  checkCollectives();
}
