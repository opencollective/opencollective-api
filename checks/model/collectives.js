import '../../server/env';

import { sequelize } from '../../server/models';

import { runCheckThenExit } from './_utils';

async function checkDeletedUsers() {
  const message = 'No USER Collective without a matching User (no auto fix)';

  const results = await sequelize.query(
    `SELECT COUNT(*) as count
     FROM "Collectives" c
     LEFT JOIN "Users" u
     ON c."id" = u."CollectiveId"
     WHERE c."type" = 'USER'
     AND c."isIncognito" IS FALSE
     AND c."deletedAt" IS NULL
     AND (u."deletedAt" IS NOT NULL or u."id" IS NULL)`,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results[0].count > 0) {
    // Not fixable
    throw new Error(message);
  }
}

async function checkActiveApprovedAtInconsistency() {
  const message = 'approvedAt and isActive are inconsistent (no auto fix)';

  const [results] = await sequelize.query(
    `
    SELECT
      COUNT(*) FILTER (WHERE "isActive" IS TRUE and "approvedAt" IS NULL) as "activeUnapproved",
      COUNT(*) FILTER (WHERE "isActive" IS NOT TRUE and "approvedAt" IS NOT NULL) as "inactiveApproved"
    FROM "Collectives"
    WHERE "deletedAt" IS NULL
    AND (
      ("isActive" IS TRUE and "approvedAt" IS NULL)
      OR ("isActive" IS NOT TRUE and "approvedAt" IS NOT NULL)
    )`,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results.activeUnapproved > 0 || results.inactiveApproved > 0) {
    throw new Error(
      `${message} (${results[0].activeUnapproved} activeUnapproved, ${results[0].inactiveApproved} inactiveApproved)`,
    );
  }
}

export async function checkCollectives() {
  await checkDeletedUsers();
  await checkActiveApprovedAtInconsistency();
}

if (!module.parent) {
  runCheckThenExit(checkCollectives);
}
