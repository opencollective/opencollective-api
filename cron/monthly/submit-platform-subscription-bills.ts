import '../../server/env';

import { Parser } from '@json2csv/plainjs';
import config from 'config';
import { compact, groupBy, sumBy } from 'lodash';
import moment from 'moment';

import activityType from '../../server/constants/activities';
import { SupportedCurrency } from '../../server/constants/currencies';
import expenseStatus from '../../server/constants/expense-status';
import expenseTypes from '../../server/constants/expense-type';
import PlatformConstants from '../../server/constants/platform';
import logger from '../../server/lib/logger';
import { notify } from '../../server/lib/notifications/email';
import { parseToBoolean } from '../../server/lib/utils';
import models, { Collective, Op, PlatformSubscription, sequelize } from '../../server/models';
import PayoutMethod, { PayoutMethodTypes } from '../../server/models/PayoutMethod';
import { Billing } from '../../server/models/PlatformSubscription';
import { runCronJob } from '../utils';

const json2csv = (data, opts = undefined) => new Parser(opts).parse(data);

const today = moment.utc();

const defaultDate = process.env.START_DATE ? moment.utc(process.env.START_DATE) : moment.utc();

const DRY = process.env.DRY;
const isProduction = config.env === 'production';

// Only run on the 1th of the month
if (isProduction && new Date().getDate() !== 1 && !process.env.OFFCYCLE) {
  console.log('OC_ENV is production and today is not the 1st of month, script aborted!');
  process.exit();
} else if (parseToBoolean(process.env.SKIP_PLATFORM_BILLING)) {
  console.log('Skipping because SKIP_PLATFORM_BILLING is set.');
  process.exit();
}

if (DRY) {
  logger.warn('Running dry, changes are not going to be persisted to the DB.');
}

export async function run(baseDate: Date | moment.Moment = defaultDate): Promise<void> {
  const momentDate = moment(baseDate);
  const billingPeriodDate = moment(momentDate).subtract(1, 'months');
  const year = billingPeriodDate.year();
  const month = billingPeriodDate.month();
  logger.info(`Submitting platform subscription bills for ${billingPeriodDate.format('MMMM/YYYY')}...`);

  const payoutMethods = groupBy(
    await models.PayoutMethod.findAll({
      where: { CollectiveId: PlatformConstants.PlatformCollectiveId, isSaved: true },
    }),
    'type',
  ) as Record<PayoutMethodTypes, PayoutMethod[]>;
  const settlementBankAccountPayoutMethod = payoutMethods[PayoutMethodTypes.BANK_ACCOUNT].find(
    pm => pm.data?.['currency'] === 'USD',
  );

  const distinctActiveSubscriptionsByCollectiveId = await PlatformSubscription.findAll({
    where: {
      period: {
        [Op.overlap]: PlatformSubscription.getBillingPeriodRange({ year, month }),
      },
    },
    attributes: [[sequelize.fn('DISTINCT', sequelize.col('CollectiveId')), 'CollectiveId']],
    order: [['CollectiveId', 'ASC']],
    raw: true,
  });
  logger.info(`Found ${distinctActiveSubscriptionsByCollectiveId.length} active subscribers...`);

  const currency = 'USD' as SupportedCurrency;
  for (let i = 0; i < distinctActiveSubscriptionsByCollectiveId.length; i++) {
    const logPrefix = `[${i + 1}/${distinctActiveSubscriptionsByCollectiveId.length}]`;
    const subscription = distinctActiveSubscriptionsByCollectiveId[i];
    const orgId = subscription.CollectiveId;
    try {
      const organization: Collective = await models.Collective.findByPk(orgId);
      logger.info(`${logPrefix} Processing subscriptions for organization ${organization.name} #${organization.id}`);
      const bill = await PlatformSubscription.calculateBilling(organization.id, { year, month });
      // Ignore bills for $0
      if (bill.totalAmount === 0) {
        logger.info(`${logPrefix} Ignoring $0 bill for organization ${organization.name} #${organization.id}`);
        continue;
      } else {
        logger.info(
          `${logPrefix} Processing bill for organization ${organization.name} #${organization.id} total: \$${bill.totalAmount / 100}`,
        );
      }

      // Check if we already billed this organization for the current billing period
      const existingExpense = await models.Expense.findOne({
        where: {
          CollectiveId: organization.id,
          type: expenseTypes.PLATFORM_BILLING,
          createdAt: {
            [Op.gte]: moment(billingPeriodDate).subtract(2, 'month').toDate(),
          },
          data: { bill: { billingPeriod: bill.billingPeriod } },
        },
      });
      if (existingExpense) {
        logger.info(
          `${logPrefix} Organization ${organization.name} #${organization.id} has already been billed for this period, skipping...`,
        );
        continue;
      }

      function subscriptionBaseItemDescription(sub: Billing['base']['subscriptions'][number]) {
        const startDate = moment.utc(sub.startDate).format('DD-MMM-YYYY');
        const endDate = moment.utc(sub.endDate).format('DD-MMM-YYYY');
        return `Base subscription ${sub.title} - ${startDate} to ${endDate}`;
      }

      const incurredAt = moment.utc(bill.billingPeriod).toDate();
      let items = compact([
        ...bill.base.subscriptions
          .filter(sub => sub.amount > 0)
          .map(sub => ({
            description: subscriptionBaseItemDescription(sub),
            amount: sub.amount,
            incurredAt: sub.endDate,
            currency,
          })),
        bill.additional.utilization.activeCollectives > 0 && {
          description: `Additional Active Collective Utilization: ${bill.additional.utilization.activeCollectives}`,
          amount: bill.additional.amounts.activeCollectives,
          incurredAt,
          currency,
        },
        bill.additional.utilization.expensesPaid > 0 && {
          description: `Additional Paid Expenses Utilization: ${bill.additional.utilization.expensesPaid}`,
          amount: bill.additional.amounts.expensesPaid,
          incurredAt,
          currency,
        },
      ]);

      const payoutMethod = await PlatformSubscription.getPreferredPlatformPayout(
        organization,
        payoutMethods,
        settlementBankAccountPayoutMethod,
      );

      if (!payoutMethod) {
        throw new Error('No Payout Method found, Open Collective Inc. needs to have at least one payout method.');
      }

      const totalAmountCharged = sumBy(items, 'amount');
      const expenseData = {
        FromCollectiveId: PlatformConstants.PlatformCollectiveId,
        lastEditedById: PlatformConstants.PlatformUserId,
        UserId: PlatformConstants.PlatformUserId,
        payeeLocation: {
          address: PlatformConstants.PlatformAddress,
          country: PlatformConstants.PlatformCountry,
        },
        PayoutMethodId: payoutMethod.id,
        amount: totalAmountCharged,
        CollectiveId: organization.id,
        currency: 'USD' as SupportedCurrency,
        description: `Platform subscription payment for ${billingPeriodDate.format('MMMM, YYYY')}`,
        incurredAt: today.toDate(),
        data: { bill },
        type: expenseTypes.PLATFORM_BILLING,
        status: expenseStatus.APPROVED,
      };

      if (DRY) {
        console.debug(`Expense:\n${JSON.stringify(expenseData, null, 2)}`);
        console.debug(`PayoutMethod: ${payoutMethod.id} - ${payoutMethod.type}`);
        console.debug(`Items:\n${json2csv(items)}\n`);
      } else {
        // Create the Expense
        const expense = await models.Expense.create(expenseData);
        // Create Expense Items
        items = items.map(i => ({
          ...i,
          ExpenseId: expense.id,
          CreatedByUserId: PlatformConstants.PlatformUserId,
        }));
        await models.ExpenseItem.bulkCreate(items);

        const platformUser = await models.User.findByPk(PlatformConstants.PlatformUserId);
        const newExpenseActivity = await expense.createActivity(activityType.COLLECTIVE_EXPENSE_CREATED, platformUser, {
          notify: false,
        });

        try {
          await PlatformSubscription.chargeExpense(expense);
        } catch (err) {
          logger.error(`Error while charging platform expense #${expense.id} to #${orgId}: ${err.message}`);
          await notify.collective(newExpenseActivity, {
            template: 'platform.billing.new.expense',
          });
        }
      }
    } catch (e) {
      logger.error(
        `Error occurred while submitting organization  platform subscription bill to #${orgId}: ${e.message}`,
      );
    }
  }
}

if (require.main === module) {
  runCronJob('submit-platform-subscription-bills', () => run(defaultDate), 23 * 60 * 60);
}
