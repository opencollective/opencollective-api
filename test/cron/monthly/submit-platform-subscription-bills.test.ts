import { expect } from 'chai';
import { pick } from 'lodash';
import moment from 'moment';
import sinon from 'sinon';

import { run } from '../../../cron/monthly/submit-platform-subscription-bills';
import { expenseTypes } from '../../../server/constants';
import { CollectiveType } from '../../../server/constants/collectives';
import { PlatformSubscriptionTiers } from '../../../server/constants/plans';
import PlatformConstants from '../../../server/constants/platform';
import models, { PlatformSubscription } from '../../../server/models';
import { PayoutMethodTypes } from '../../../server/models/PayoutMethod';
import {
  fakeCollective,
  fakeConnectedAccount,
  fakeHost,
  fakePayoutMethod,
  fakeUser,
  randStr,
} from '../../test-helpers/fake-data';
import { resetTestDB } from '../../utils';

describe('submit-platform-subscription-bills', () => {
  const date = moment.utc('2023-10-09T10:00:00Z');
  const sandbox = sinon.createSandbox();

  let organizations;
  before(async () => {
    await resetTestDB();
    const user = await fakeUser({ id: PlatformConstants.PlatformUserId }, { slug: 'ofitech-admin' });
    const oc = await fakeHost({
      id: PlatformConstants.PlatformCollectiveId,
      slug: randStr('platform-'),
      CreatedByUserId: user.id,
    });
    await fakeConnectedAccount({ CollectiveId: oc.id, service: 'stripe' });
    const payoutProto = {
      data: {
        details: {},
        type: 'IBAN',
        accountHolderName: 'OpenCollective Inc.',
        currency: 'USD',
      },
      CollectiveId: oc.id,
      type: PayoutMethodTypes.BANK_ACCOUNT,
    };
    await fakePayoutMethod({
      ...payoutProto,
      id: 2955,
      isSaved: true,
    });

    // ocStripePayoutMethod = (await oc.getPayoutMethods()).find(pm => pm.type === PayoutMethodTypes.STRIPE);
    organizations = [];
    for (let i = 0; i < 5; i++) {
      organizations.push(await fakeCollective({ type: CollectiveType.ORGANIZATION, isActive: true }));
    }

    const plans = [
      PlatformSubscriptionTiers[0],
      PlatformSubscriptionTiers[1],
      PlatformSubscriptionTiers[2],
      PlatformSubscriptionTiers[3],
      PlatformSubscriptionTiers[4],
    ];
    for (const org of organizations) {
      const i = organizations.indexOf(org);
      await PlatformSubscription.createSubscription(org, moment(date).subtract(2, 'month').toDate(), plans[i], user);
    }

    const calculateUtilizationStub = sandbox.stub(PlatformSubscription, 'calculateUtilization');
    const utilizations = [
      { activeCollectives: 0, expensesPaid: 5 },
      { activeCollectives: 5, expensesPaid: 2 },
      { activeCollectives: 4, expensesPaid: 55 },
      { activeCollectives: 20, expensesPaid: 120 },
      { activeCollectives: 30, expensesPaid: 230 },
    ];
    organizations.forEach((org, i) => {
      calculateUtilizationStub.withArgs(org.id).resolves(utilizations[i]);
    });
  });

  after(() => {
    sandbox.restore();
  });

  it('should run without errors', async () => {
    await run(date);
  });

  it('should omit bills for $0', async () => {
    const expenses = await models.Expense.findAll({
      where: { CollectiveId: organizations[0].id, type: expenseTypes.PLATFORM_BILLING },
    });

    expect(expenses).to.have.length(0);
  });

  it('should submit Expenses for active subscriptions', async () => {
    const expenses = await models.Expense.findAll({
      where: { type: expenseTypes.PLATFORM_BILLING },
      order: [['id', 'DESC']],
    });

    expect(expenses).to.have.length(4);
    const expensesTable = expenses.map(e =>
      pick(e.toJSON(), ['id', 'type', 'CollectiveId', 'description', 'amount', 'currency']),
    );
    expect(expensesTable).to.matchTableSnapshot();

    const items = await models.ExpenseItem.findAll({
      where: { ExpenseId: expenses.map(e => e.id) },
      order: [['id', 'DESC']],
    });

    const itemsTable = items.map(i => pick(i.toJSON(), ['ExpenseId', 'description', 'amount', 'currency']));
    expect(itemsTable).to.matchTableSnapshot();
  });

  it('should not bill organizations twice', async () => {
    await run(date);

    const expenses = await models.Expense.findAll({
      where: { CollectiveId: organizations[1].id, type: expenseTypes.PLATFORM_BILLING },
    });

    expect(expenses).to.have.length(1);
  });
});
