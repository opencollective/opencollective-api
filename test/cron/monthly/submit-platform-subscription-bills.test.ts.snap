// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submit-platform-subscription-bills should submit Expenses for active subscriptions 1`] = `
"
| id | type             | CollectiveId | description                                       | amount | currency |
| -- | ---------------- | ------------ | ------------------------------------------------- | ------ | -------- |
| 4  | PLATFORM_BILLING | 22           | Platform subscription payment for September, 2023 | 69500  | USD      |
| 3  | PLATFORM_BILLING | 18           | Platform subscription payment for September, 2023 | 21000  | USD      |
| 2  | PLATFORM_BILLING | 14           | Platform subscription payment for September, 2023 | 4500   | USD      |
| 1  | PLATFORM_BILLING | 10           | Platform subscription payment for September, 2023 | 4000   | USD      |"
`;

exports[`submit-platform-subscription-bills should submit Expenses for active subscriptions 2`] = `
"
| ExpenseId | description                                  | amount | currency |
| --------- | -------------------------------------------- | ------ | -------- |
| 4         | Additional Paid Expenses Utilization: 180    | 27000  | USD      |
| 4         | Additional Active Collective Utilization: 25 | 37500  | USD      |
| 4         | Base Subscription Amount                     | 5000   | USD      |
| 3         | Additional Paid Expenses Utilization: 20     | 2000   | USD      |
| 3         | Additional Active Collective Utilization: 10 | 10000  | USD      |
| 3         | Base Subscription Amount                     | 9000   | USD      |
| 2         | Additional Paid Expenses Utilization: 5      | 500    | USD      |
| 2         | Base Subscription Amount                     | 4000   | USD      |
| 1         | Additional Active Collective Utilization: 2  | 2000   | USD      |
| 1         | Base Subscription Amount                     | 2000   | USD      |"
`;
